programs/lz4io.c

```diff
@@ -1362,47 +1398,204 @@ const char* LZ4IO_blockTypeID(int sizeID, int blockMode, char* buffer)
+    sprintf(buf, "%.2Lf%c", size, units[i]);
+    return buf;
+}
 
-int LZ4IO_displayCompressedFilesInfo(const char** inFileNames, size_t ifnIdx)
+/* Get filename without path prefix */
+static const char* LZ4IO_baseName(const char* input_filename) {
+  const char* b = strrchr(input_filename, '/');
+  if (!b) b = strrchr(input_filename, '\\');
+  return b ? b + 1 : b;
+}
+
+/* Report frame/s information in verbose mode.
+ * Will populate file info with fileName and contentSize where applicable.
+ * - TODO :
+ *  + report nb of blocks, hence max. possible decompressed size (when not reported in header)
+ */
+static LZ4IO_infoResult
+LZ4IO_getCompressedFileInfo(LZ4IO_cFileInfo_t* cfinfo, const char* input_filename)
 {
-    int result = 0;
-    size_t idx;
+    LZ4IO_infoResult result = LZ4IO_format_not_known;  /* default result (error) */
+    unsigned char buffer[LZ4F_HEADER_SIZE_MAX];
+    FILE* const finput = LZ4IO_openSrcFile(input_filename);
+    cfinfo->fileSize = UTIL_getFileSize(input_filename);
 
-    DISPLAY("%5s %20s %20s %10s %7s  %s\n",
-        "Block", "Compressed", "Uncompressed", "Ratio", "Check", "Filename");
 
-    for (idx=0; idx<ifnIdx; idx++) {
+    while(!feof(finput)){
+      { LZ4IO_frameInfo_t frameInfo = LZ4IO_INIT_FRAMEINFO;
+        unsigned magicNumber;
+        /* Get MagicNumber */
+        size_t nbReadBytes = fread(buffer, 1, MAGICNUMBER_SIZE, finput);
+        if (nbReadBytes==0) { break; }   /* EOF */
+        result = LZ4IO_format_not_known;  /* default result (error) */
+        if (nbReadBytes != MAGICNUMBER_SIZE)
+          EXM_THROW(40, "Unrecognized header : Magic Number unreadable");
+        magicNumber = LZ4IO_readLE32(buffer);   /* Little Endian format */
+        if (LZ4IO_isSkippableMagicNumber(magicNumber))
+            magicNumber = LZ4IO_SKIPPABLE0;   /* fold skippable magic numbers */
+
+        switch(magicNumber)
+        {
+        case LZ4IO_MAGICNUMBER:
+            if(cfinfo->frameSummary.frameType != lz4Frame) cfinfo->eqFrameTypes = 0;
+            /* Get frame info */
+            { const size_t readBytes = fread(buffer+MAGICNUMBER_SIZE, 1, LZ4F_HEADER_SIZE_MIN-MAGICNUMBER_SIZE, finput);
+            if (!readBytes || ferror(finput)) EXM_THROW(71, "Error reading %s", input_filename); }
+            {  size_t hSize = LZ4F_headerSize(&buffer, LZ4F_HEADER_SIZE_MIN);
+            if(!LZ4F_isError(hSize)){
+              if(hSize > (LZ4F_HEADER_SIZE_MIN + MAGICNUMBER_SIZE)){
+                /* We've already read LZ4F_HEADER_SIZE_MIN so read any extra until hSize*/
+                const size_t readBytes = fread(buffer+LZ4F_HEADER_SIZE_MIN, 1, hSize-LZ4F_HEADER_SIZE_MIN, finput);
+                if (!readBytes || ferror(finput)) EXM_THROW(72, "Error reading %s", input_filename);
+              }
+              /* Create decompression context */
+              { LZ4F_dctx* dctx;
+                if (!LZ4F_isError(LZ4F_createDecompressionContext(&dctx, LZ4F_VERSION))) {
```

Comment:
On this line, `dctx` is allocated,
but I don't see the corresponding `free` operation ?

Reply:
Added free() immediately after it becomes unneeded.